### PR TITLE
Introduce test protocol for primitive tests

### PR DESCRIPTION
**Pull request by kuujo**
This PR introduces a new local test `ProxyProtocol` for use in primitive tests. The test protocol uses only threading to simulate a distributed cluster/messaging and uses a simple shared, single threaded protocol for managing services in tests. This ensures that primitive tests are deterministic. Protocol tests should be solely responsible for testing the correctness of the protocols themselves.

### CODE REVIEW

**Comment by kuujo** [comment]
BTW much of this PR is deleting Raft/primary-backup tests.

**Comment by johnou** [comment]
Does this supersede
 https://github.com/atomix/atomix/pull/819?

**Comment by kuujo** [comment]
I think so. We still need to do further work on parallel tests, but this makes #819 unnecessary. It also feels much better since it makes primitive tests deterministic.

File: core/src/test/java/io/atomix/core/AbstractPrimitiveTest.java [hunk 1]
```diff
@@ -15,101 +15,74 @@
...
   @After
   public void teardownTest() throws Exception {
-    List<CompletableFuture<Void>> futures = clients.stream().map(Atomix::stop).collect(Collectors.toList());
+    List<CompletableFuture<Void>> futures = members.stream().map(Atomix::stop).collect(Collectors.toList());
     try {
       CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()])).get(30, TimeUnit.SECONDS);
     } catch (Exception e) {
       // Do nothing
     }
+    protocol.close();
```
[the remark this answers is not in the document]
Comment (kuujo) [reply core/src/test/java/io/atomix/core/AbstractPrimitiveTest.java]: I guess it doesn't matter since the exception is ignored anyways, but sure

**Pull request merged**